File: in_Java/0144 Binary Tree Preorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if(root == null) return res;
        helper(root,res);
        return res;
    }
    private void helper(TreeNode root, List<Integer> res){
        if(root == null) return;
        res.add(root.val);
        helper(root.left,res);
        helper(root.right,res);
    }
}


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        Stack<TreeNode> st = new Stack<>();
        if(root == null) return res;
        st.push(root);
        TreeNode cur;
        while(!st.isEmpty()){
            cur = st.pop();
            res.add(cur.val);
            if(cur.right != null) st.push(cur.right);
            if(cur.left != null) st.push(cur.left);
        }
        return res;
    }
}

File: in_Python/0144 Binary Tree Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        stack = []
        if root == None:
            return stack
        stack.append(root)
        while len(stack) != 0:
            cur_node = stack.pop()
            res.append(cur_node.val)
            if cur_node.right != None:
                stack.append(cur_node.right)
            if cur_node.left != None:
                stack.append(cur_node.left)            
        return res